frontend/catalyst/jax_primitives.py

```diff
@@ -209,12 +210,22 @@ def _func_def_impl(ctx, *args, call_jaxpr, fn, call=True):  # pragma: no cover
     raise NotImplementedError()
 
 
-def _func_symbol_lowering(ctx, fn_name, call_jaxpr):
+def _func_def_lowering(ctx, fn, call_jaxpr) -> str:
     """Create a func::FuncOp from JAXPR."""
     if isinstance(call_jaxpr, core.Jaxpr):
         call_jaxpr = core.ClosedJaxpr(call_jaxpr, ())
-    symbol_name = mlir.lower_jaxpr_to_fun(ctx, fn_name, call_jaxpr, tuple()).name.value
-    return symbol_name
+    func_op = mlir.lower_jaxpr_to_fun(ctx, fn.__name__, call_jaxpr, tuple())
+
+    if isinstance(fn, pennylane_QNode):
+        func_op.attributes["qnode"] = ir.UnitAttr.get()
+        # "best", the default option in PennyLane, chooses backprop on the device
+        # if supported and parameter-shift otherwise. Emulating the same behaviour
+        # would require generating code to query the device.
+        # For simplicity, Catalyst instead defaults to finite-diff.
+        diff_method = fn.diff_method if fn.diff_method != "best" else "finite-diff"
+        func_op.attributes["diff_method"] = ir.StringAttr.get(diff_method)
```

Comment:
@dime10 should we make the default diff method `parameter-shift` for now?

Reply:
Can't we check the device here and if it's lightning choose adjoint, else parameter-shift?

Finite-diff has the best compatibility however and we don't have a way to fall back on it if the other two don't work.

Reply:
> Can't we check the device here and if it's lightning choose adjoint, else parameter-shift?

This is a good first-order approach, however care would need to be taken at the circuit level, since not all ops and measurements will support parameter-shift/adjoint.

In terms of making choice `adjoint/parameter-shift/finite-diff` in that order, what would need to be done/what would Catalyst need to know?

@dime10 something to also add - in Python PL, `diff-method="param-shift"` still  determines the differentiation method _per parameter_. That is, if a circuit contains X ops that support param-shift, and 1 op that does not, we use param-shift on all but the one op that does not, and use finite diff for that one.

Here is the function we use: https://github.com/PennyLaneAI/pennylane/blob/e102fe2ad134b8834a27dc2cad08c4b6de2f62e0/pennylane/gradients/gradient_transform.py#L174

We could probably call this function directly in Catalyst

Reply:
Yup that's what I mean, we don't have a way to fall back on other methods in Catalyst right now, that's why I still prefer fd default at the moment.

Per-operation might be quite complex to add, without the compiler knowing what each of the operations mean (right now they are generic).

Reply:
This ties into the bigger picture I think of supporting general shift rules in catalyst.

> Per-operation might be quite complex to add, without the compiler knowing what each of the operations mean (right now they are generic).

I think the compiler doesn't need to know explicitly what each of the operations are; potentially all it needs to know (or be able to compute) is `[shifts, coefficients, multipliers]` for each op. If an op has frequencies/a generator, the param shifts can be computed, otherwise finite diff can be used (this data format is general enough to cover both param shift _and_ finite diff, as well as gradients of channels).

yep, would be good to understand what we would need to change to support the same functionality python PL currently has.

Reply:
Alright, so within the context of this PR, is the consensus to stick with finite diff as the default?

Reply:
From my side that would be acceptable for this PR 👍